design-docs/CacheControl.md

```diff
@@ -0,0 +1,138 @@
+# CacheControl
+
+## Context
+
+In a traditional HTTP API, the semantics of CacheControl are well-defined. Mozilla for an example has a [very detailed documentation about the different cache headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching).
+
+In a GraphQL API, HTTP caching doesn't work as well because the data might overlap between two queries. Because we want the data to be de-duplicated to have a single source of truth, HTTP caching only get us that far.
+
+This document is a working document to add CacheControl to the normalized cache
+
+## Server cache control 
+
+Apollo Server has a [`@cacheControl` directive](https://www.apollographql.com/docs/apollo-server/performance/caching/) for cache control.
+
+```graphql
+# A session can be cached 10min
+type Session @cacheControl(maxAge: 600) {
+  # But dates might change often so they have a 30s expiration
+  startDate: Date! @cacheControl(maxAge: 30)
+  endDate: Date! @cacheControl(maxAge: 30)
+  title: String!
+  speaker: Speaker!
+}
+
+# Speaker details do not change often so we cache them for 1h
+type Speaker @cacheControl(maxAge: 3600) {
+  name: String!
+  bio: String!
+  twitter: String
+}
+```
+
+The `@cacheControl` directives are analyzed and then exposed in a `Cache-Control` HTTP header. Because there is only one `Cache-Control` HTTP header for the whole query, the most restrictive `maxAge` is used. A query like this:
+
+```graphql
+{
+  session(id: "42") {
+    # Because of startDate, the whole query will only be cached 30s
+    startDate
+    title
+    speaker {
+      name
+      bio
+    }
+  }
+}
+```
+
+will only be cached 30s because it has a field that changes often (`startDate`)
+
+## Client cache control
+
+The same server-side rules could be added client side:
+
+```graphql
+extend type Session @cacheControl(maxAge: 600) {
+  startDate: Date! @cacheControl(maxAge: 30)
+  endDate: Date! @cacheControl(maxAge: 30)
+}
+
+extend type Speaker @cacheControl(maxAge: 3600) 
+```
+
+Or if introspection ever supports directive usages, it could be retrieved (and cached) at runtime.
+
+## Storing the cache control information
+
+We can think of several fields to store (loosely related to their equivalent HTTP definitions):
+
+- Date: the date the data originated (details TBC: created vs sent vs receive vs other?)
+- MaxAge: the maximum age before the data becomes stale
+- StaleWhileRevalidate
+- ...
+
+Date sounds required in all cases. MaxAge is required if we want to store that information when coming from the server but could be spared if the rules are defined on the client side.
```

Comment:
> Date sounds required in all cases.
Ideally date is optional. If server does not send it, client can generate the "received" date.